code/datums/components/z_parallax.dm

```diff
@@ -0,0 +1,73 @@
+/**
+ * Component that hooks into the client, listens for COMSIG_MOVABLE_Z_CHANGED, and depending on whether or not the
+ * Z-level has ZTRAIT_NOPARALLAX enabled, disable or reenable parallax.
+ */
+
+/datum/component/zparallax
+	dupe_mode = COMPONENT_DUPE_UNIQUE
+
+	/// If TRUE, disable parallax checking alltogether until we change Z-levels again.
+	var/caching = FALSE
+
+	var/client/tracked
+	var/mob/client_mob
+
+/datum/component/zparallax/Initialize(client/tracked)
+	. = ..()
+	if(!istype(tracked))
+		stack_trace("Component zparallax has been initialized outside of a client. Deleting.")
+		return COMPONENT_INCOMPATIBLE
+
+	src.tracked = tracked
+	client_mob = tracked.mob
+
+/datum/component/zparallax/RegisterWithParent()
+	RegisterSignal(client_mob, COMSIG_MOB_LOGOUT, .proc/mob_change)
+	RegisterSignal(client_mob, COMSIG_MOVABLE_Z_CHANGED, .proc/ztrait_checks)
+	RegisterSignal(client_mob, COMSIG_MOB_LOGIN, .proc/refresh_client)
+
+/datum/component/zparallax/UnregisterFromParent()
+	unregister_signals()
+
+	tracked = null
+	client_mob = null
+
+/datum/component/zparallax/proc/unregister_signals()
+	if(!client_mob)
+		return
+
+	UnregisterSignal(client_mob, list(COMSIG_MOB_LOGOUT, COMSIG_MOVABLE_Z_CHANGED))
+
+/datum/component/zparallax/proc/refresh_client()
+	tracked = client_mob.client
+
+/datum/component/zparallax/proc/mob_change()
+	SIGNAL_HANDLER
+
+	if(client_mob.key)
+		return
+
+	unregister_signals()
+
+	client_mob = tracked.mob
+
+	RegisterSignal(client_mob, COMSIG_MOB_LOGOUT, .proc/mob_change)
+	RegisterSignal(client_mob, COMSIG_MOVABLE_Z_CHANGED, .proc/ztrait_checks)
+	RegisterSignal(client_mob, COMSIG_MOB_LOGIN, .proc/refresh_client)
+
+/datum/component/zparallax/proc/ztrait_checks()
+	SIGNAL_HANDLER
+
+	client_mob = tracked.movingmob
+	var/parallax = SSmapping.level_trait(client_mob.z, ZTRAIT_NOPARALLAX)
+	var/datum/hud/hud = client_mob.hud_used
+
+	if(tracked.prefs.read_preference(/datum/preference/choiced/parallax) == PARALLAX_DISABLE)
+		return
+
+	if(!parallax)
+		hud.create_parallax(client_mob)
+		return
+	else
+		hud.remove_parallax(client_mob)
+		return
```

Comment:
It might be better to stick this logic inside `create_parallax` directly, the parallax checking stuff I mean, and have moving a z level just call `update_parallax_pref`.

That way any rouge calls to `create_parallax` won't break our optimization.

Reply:
Actually why do you even need to call remove/create parallax like this? don't you already have a check that has this effect inside `apply_parallax_pref`?